### ISSUE TITLE
Is it possible to leave stderr messages from the compiler on stderr and not redirect them to stdout?

### ISSUE THREAD

**Issue by tpimh** [ninja-build/ninja#1537]
I understand that default ninja behavior is to only print to stdout, but I wonder if it's possible to implement this as an option like done in this branch: https://github.com/dendy/ninja/tree/stderr

**Comment by jhasse** [comment]
What do you need it for?

**Comment by tpimh** [comment]
I would like to separate build output into too files: `build.log` for informational output and `build.err` for warnings and errors. Redirecting stdout to one file and stderr to another works for GNU Make, but not for Ninja.

**Comment by jhasse** [comment]
You won't know which warning or error belongs to which build output line. What's the reason behind the two files?

**Comment by tpimh** [comment]
Mostly just warning detection. If there are warnings, they should be either fixed or disabled. The warning usually contains source filename and line number, so there is no need to find corresponding lines of ninja's own output.

**Comment by jhasse** [comment]
I see. Would #480 be enough for your use-case?

The reason I'm asking is: We're trying to keep the number of options as low as possible.

**Comment by tpimh** [comment]
I'm not sure how could possibly be the solution for my problem. Probably there may be a very hacky solution with NINJA_STATUS variable, but I can't think of one right now.
On the other hand if split stderr is implemented, `--quiet` that disables all ninja's own output and only prints compiler output would be as easy as `ninja --split-sderr >/dev/null`.

**Comment by jhasse** [comment]
> I'm not sure how that could possibly be the solution for my problem.

You would run `ninja --quiet` and could detect all the warnings. (I'm aware that `--quiet` isn't implemented yet)

**Comment by tpimh** [comment]
But in this case I will not get any ninja's own output. I would like to keep both, but split them.

**Comment by jhasse** [comment]
Why?

**Comment by tpimh** [comment]
In current build script I print a dot for every line on stdout and an x for every line on stderr. Saving both stdout and stderr into separate files. This keeps the CI from stopping because of the lack of output, gives a pretty visual output that helps to detect new warnings and errors and provides separate readable build logs with warning descriptions. This works for other build systems, but not for ninja as there is no way to tell if a line came from stderr or stdout. This is how I did it: https://github.com/tpimh/nenuzhnix-build/commit/722ece30d2b3801b7871ae091223e261c38e671d (still to be rewritten properly)

**Comment by jhasse** [comment]
Set `NINJA_STATUS="[ninja][%f/%t] "` and then you `grep` for `"^[ninja]"` in `process_out` to see if it's ninja output or compiler output.

**Comment by tpimh** [comment]
Yes, can be done this way. Thanks!

**Comment by tpimh** [comment]
This is my script for restoring justice over stdout and stderr:

```bash
filter() {
  while read -r
  do
    echo $REPLY | grep '^\[ninja\]' | sed 's/\[ninja\]//'
    echo $REPLY | grep -v '^\[ninja\]' >&2
  done
}

NINJA_STATUS="[ninja][%f/%t] " ninja | filter
```
It may be a bit slow, but it works for me. Feel free to use and improve.

**Comment by subdiff** [comment]
Same for QtCreator.

Just a simple script to wrap ninja with stdout > stderr is enough. But it would be nice if it would work out of the box.

**Comment by kloczek** [comment]
Can you share what and how you are actually using as the wrpaper?

**Comment by tpimh** [comment]
I have [my own ninja wrapper](https://github.com/tpimh/nenuzhnix-build/blob/docker/root/nenuzhnix-tools/ninjahelper). It's really bad, but it works. Written in POSIX shell, could be much easier in bash, but it's unfortunately not an option for me.

**Comment by subdiff** [comment]
```
$ cat /usr/bin/ninja-stderr

#!/bin/sh
/usr/bin/ninja "$@" 1>&2
```

**Comment by sglass68** [comment]
Is there any chance this might be 'fixed' in ninja itself? If ninja considers all of its output to be errors, then perhaps it should just move to stderr. If it considers some of it to be informational, not indicating warnings/failures, then that part should go to stdout IMO.

**Comment by jhasse** [comment]
@sjg20 #1210 would allow this to be implemented in a cleaner way.

**Comment by dendy** [comment]
Hello, I am the author of the patch for this issue mentioned in the first comment: https://github.com/dendy/ninja/tree/stderr

Patch is currently updated with Linux/Mac/Windows support and rebased onto master branch.

I consider this issue as critical. Without separation of the stdout and stderr usage of the ninja is limited mostly to the CLI. IDEs and tools which expect the proper separation of output, like it was in the Makefiles, will not work with ninja at all.

Here is the thing. Decision where to put an output should be made by the executing command, not the ninja itself. If some command decides to put portion of output to the stderr, that should be possible and properly pop up at the ninja's end in the stderr. Generally speaking stdout and stderr supposed to be completely independent, and it is up to the command itself to ensure that both streams are contiguous. And whoever is parsing the output should expect that stdout and stderr are not connected together.

Now, I understand why ninja is combining both outputs, to keep them in sync with printing the execution command, so it is easier to see whole output in the solid chunk. Although such approach has it is own benefits in some cases, it is not the only one that is used in the build environments.

Particularly using ninja from IDE requires to parse stderr. Not combined stdout and stderr, but stderr alone. Each line printed by the executing command into stderr is considered as a separate issue. The executing command arguments usually don't matter in the IDE context, and even if they do there are other ways to connect errors and command line, rather than eliminating the stderr completely. In worst case it will be the same experience as with Makefiles.

I strongly believe that this feature is absolutely necessary. It should be up to the user, which method is preferable in particular case. It could be environment variable, command line argument, or some extra parameter in the build.ninja, or combination of those. But user should have control over this, whether output of the stdout and stderr should be merged together or not.

For the 5 good years I never use combined output in any of my projects, either they are built from the command line or from IDE. And I am totally happy with the results.

**Comment by RedDwarf69** [comment]
FWIW ninja is corrupting the output from Doxygen (via CMake's `doxygen_add_docs()`). The issue doesn't happen when using GNU Make (or Doxygen directly).

This is a showstopper to me since it breaks my Jenkins builds because the Warnings Next Generation plugin starts to search for files that don't exist. So to get my builds working I need to get back to GNU Make.

**Comment by RedDwarf69** [comment]
Actually I am back to using ninja. By prefixing my commands with `stdbuf -oL -eL` (from GNU coreutils) everything works as it should.

**Comment by sjg20** [comment]
@jhasse how about an option to handle stdout and stderr separately (I would say correctly)? Warnings and errors should go to stderr and that will allow IDEs to work.

**Comment by jhasse** [comment]
If this was to be fixed, it shouldn't be an option, but always handle stdout and stderr correctly. If someone wants the old (current) behavior, he has the "option" of using `2>&1`.

**Comment by kloczek** [comment]
So what is the curent state of the not redirecting (at least on demmand using some commandline switch) stderr to stdout?
And/or is it any plan to accept some patch which would provide such functional;ity?

Splitting stderr is very usefull on processing only warnigs/errors and current ninja behaviour makes such processing at least way more difficult.

**Comment by jhasse** [comment]
As I said, the patch would have to always split the two, no commandline switch please.

`ninja-with-patch 2>&1` would need to behave *exactly* like `old-ninja`. Getting this right is tricky and for a patch to be accepted it would need some unit / integration testing (see `misc/output_test.py`). Also it should provide some benchmarking that this doesn't impact  performance too much for jobs that print out A LOT of mixed stdout and stderr messages.

**Comment by dendy** [comment]
My patch on review could be used as a base for proceed with proper implementation. Generally speaking, stdout and stderr streams are independent, so if executed process does not care about ending data with the line break or flushing the pipe (which is totally normal) then attempt to combine such unfinished blocks of text will result in mixed content, while separating them will produce correct out and err texts, but desynced from each other.

Have to look at the output_test.py, I assume test should follow some "correct" way of writing into out/err streams in order to make validation possible.

**Comment by dendy** [comment]
Just ran the test, it passes. But it is simplistic and does not test stderr behavior.

Here is the pseudo code of the process, where behavior will be deviated from current (with stdout and stderr combined):

out.write('a'); out.flush();
err.write('b'); err.flush();
out.write('c'); out.flush();

Since in current implementation both stderr and stdout streams are connected to the write side of the shared output pipe, the result will merged at the process side and will be next: 'abc'

When stderr and stdout streams are connected to the separate pipes, when time comes to read from those pipes on the ninja side — there is no way to understand the sequence how data was placed into pipes in first place. I.e. out pipe will contain 'ac' and error pipe will contain 'b'. Reading them sequentially in any order would produce either: 'ac' then 'b' or: 'b' then 'ac'.

This is totally normal and expected. Output and error stream are not related by design to each other. Contiguous content is guaranteed only within the same stream. So generally intent to match 2>&1 to the original combined output cannot be fulfilled. Stream separation assumes this change of the behavior.

So far I see the only way to preserve original behavior is to have a runtime flag to alter between these two modes.

**Comment by dendy** [comment]
The described change is not a violation of current ninja behavior. Particularly because such behavior of ordering stdout and stderr text blocks cannot be controlled by ninja. It is up to each running task how to group the texts. In general case tasks should not care about issues of mixing data between streams, only about contiguous data within single stream: stdout or stderr. For example, writing into those streams simultaneously from the separate threads is a valid behavior.

I didn't find explicit mentioning about combining stdout and stderr by ninja, manual describes only this (https://ninja-build.org/manual.html):

> Command output is always buffered. This means commands running in parallel don’t interleave their output, and when a command fails we can print its failure output next to the full command line that produced the failure.

Potential change of the order of the text blocks between streams does not violate the above policy. Whole output of the single task will be preserved.

**Comment by jhasse** [comment]
> Just ran the test, it passes. But it is simplistic and does not test stderr behavior.

Yes, it would need to be extended. Just wanted to point out where the test should go.

> So far I see the only way to preserve original behavior is to have a runtime flag to alter between these two modes.

You would have to log whenever a switch between stdout and stderr appeared and then replicate that when printing the command's output.

**Comment by dendy** [comment]
There is no such switch between stdout and stderr. On the process side shared pipe's write end is duped to the fds 1 and 2, effectively redirecting stdout and stderr streams into that pipe. There is no way to tell which stream process attempts to write to. My patch splits that shared pipe into two, thus write ends of those pipes duped into fds 1 and 2 respectively. Still there is no way to tell how process is writing into those pipes. Even if pipes are flushed at the process side that does not mean those events will appear in the same order and buffer size at the ninja side.

Generally speaking currently ninja is messing up the content of the stdout and stderr, depending how process writes data output might appear as a byte mess.

Disconnected output of the streams is a normal behavior though and it is expected to be that way by processing each stream content separately.

**Comment by jhasse** [comment]
> Still there is no way to tell how process is writing into those pipes. Even if pipes are flushed at the process side that does not mean those events will appear in the same order and buffer size at the ninja side.

Why not? We're basically using the same mechanisms a terminal emulator is using.

**Comment by dendy** [comment]
That is right, that is why generally terminals cannot properly print stdout and stderr together. They might try to detect line breaks in the streams to avoid interleaved output by buffering and delaying the data. But that approach is based on pure guess. And in any case text blocks order is not guaranteed, because that was no such order in first place.

The best thing ninja can do is to buffer the whole stream contents of the process and then output them as a single chunks: first stdout, then stderr, or vice versa. Basically the same thing it is doing now.

**Comment by jhasse** [comment]
In my terminal your example still prints abc.

> [...] because that was no such order in first place.

There was an order of the flush calls.

> [...] first stdout, then stderr, or vice versa. Basically the same thing it is doing now.

??? That's not what ninja is doing now.

**Comment by dendy** [comment]
> In my terminal your example still prints abc.

It depends on race between writer and reader. If reader is busy then whole output will come as a single payload. If writer is busy then reader might do partial read. There is no way for the reader to see in how many chunks data was written, nor this is a design of the streams in first place.

> ??? That's not what ninja is doing now.

I mean ninja is doing the same with a single buffer. After separating stdout and stderr it will be two solid buffers.

**Comment by dendy** [comment]
Here is the quick example I just wrote:

print-abc.py - prints abc with delay between chars. ac goes to stdout, b goes to stderr.
```
#!/usr/bin/env python3

import sys
import time

delay = float(sys.argv[1]) * 0.001 if len(sys.argv) > 1 else 0
def sleep(secs):
	time.sleep(secs * delay)

sleep(0.5)

print('a')
sys.stdout.flush()
sleep(1)

print('b', file=sys.stderr)
sys.stderr.flush()
sleep(1)

print('c')
sys.stdout.flush()

sleep(0.5)
```

print-wrapper.py - runs print-abc.py and reads stdout and stderr.
```
#!/usr/bin/env python3

import subprocess
import threading
import sys

p = subprocess.Popen(['./print-abc.py'] + sys.argv[1:],
	stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)

def run(name, stream):
	def t(name, stream):
		for line in stream:
			print(f'{name} ({line.strip()})')
	threading.Thread(target=t, args=(name, stream,)).start()

run('out', p.stdout)
run('err', p.stderr)

while True:
	try:
		if p.wait() != None: break
	except subprocess.TimeoutExpired:
		pass
```

Run without delay between chars - it prints acb or bac at random:
```
$ ./print-wrapper.py 0
out (a)
out (c)
err (b)
$ ./print-wrapper.py 0
err (b)
out (a)
out (c)
```

Run with delay between chars - it prints abc:
```
$ ./print-wrapper.py 1
out (a)
err (b)
out (c)
```

**Comment by jhasse** [comment]
Who are "writer" and "reader"? Sorry I can't follow.

To sum up my point:

Currently calling

```python
#!/usr/bin/env python3

import sys

sys.stdout.write('a'); sys.stdout.flush();
sys.stderr.write('b'); sys.stderr.flush();
sys.stdout.write('c'); sys.stdout.flush();
```

from ninja will "print" (meaning that I see that on my terminal) `abc`. That behavior mustn't change.

**Comment by dendy** [comment]
You are getting abc just by luck. Any microfreeze in the terminal will lead to race lose, so output will be bac or acb at random.

**Comment by jhasse** [comment]
I think your print-wrapper.py doesn't work, because it's using threading. Ninja doesn't. Also `for line in stream:` isn't the correct way to check for a flush event.

**Comment by jhasse** [comment]
I don't see what my terminal has to do with it given that it's ninja who reads the output.

**Comment by dendy** [comment]
It does not matter whether it is threading or polling. Order of writes into different pipes does not determine order of events in read and size of data available at the moment of read. Any delay in the reader will lead to the case when all written data will be available at once.

**Comment by jhasse** [comment]
Isn't this due to pipe buffering?

**Comment by dendy** [comment]
Yes, that is because of the buffering. Other type of devices might provide file descriptors with more granular read/write protocols and fcntl API. But pipes do buffering as a part of their streaming design.

**Comment by dendy** [comment]
Btw, I don't know how terminal apps implemented, but as long as they don't care about how to differentiate stdout/stderr text blocks printed on the screen they could simply dup to the fds 1 and 2 after fork (like ninja does currently), so result will be always "abc" for the simple test above. I.e. such terminals cannot be used as a testing tool.

So that is approach 1, when you got sequential output from stdout and stderr together, but cannot separate the content by streams.

Case 2 is when using two pipes, but the content of those is buffered and reader cannot know the timings of write occurrences.

Good thing is that should not be an issue for the ninja. If the process is writing into both stdout and stderr, the content for those streams are independent by design of such process, ninja output should not care about ordering it, because running process does not care about the order in first place.

stdout and stderr is not the tool for colorizing the output or something, there are other ways of formatting the text. stdout and stderr is the tool to filter the output and deliver it to the different recipient on the reader side. Those recipients supposed to be standalone from each other. The only requirements for the writer is to ensure that within the single stream content is contiguous.

The initial requirement of matching the output to the 2>&1 is not critical. Unless some job parser was implemented with a bug and tested process output only through the ninja output. Yes, this will be known issue, but importance of it is overestimated from my point of view.

This is the same type of change in the tool when API didn't define behavior, and in later release behavior was fixed, revealing bugs in the client apps. Sure it makes sense to list that in the release notes, but should not be a showstopper. Maybe even change the major version of the app to highlight such change and let client apps time to fix their bugs, allowing not to upgrade to the newer version automatically.

**Comment by jhasse** [comment]
What about using `O_DIRECT` with [pipe2](https://man7.org/linux/man-pages/man2/pipe.2.html)?

Also asuming that streams are flushed on a newline, reading until `\n` will probably be right most of the time (works with your print-abc.py example).

There are many programs which do care about the ordering so this *is* a showstopper, sorry.

**Comment by jonesmz** [comment]
Which programs, specifically?

If they exist, they have always been wrong to rely on the order of output with a mixed stdout and stderr.

**Comment by dendy** [comment]
O_DIRECT is cannot be used, if you read the documentation about it you will see caveats about internal buffer size. Since the pipe is going to be used by arbitrary processes they will not follow any rules about writing policies into stdout or stderr. It is totally normal case to do writes in multiple chunks like:
```
sys.stdout.write("hello ")
sys.stdout.write("world")
sys.stdout.write("\n")
```
Within those calls actual write could appear at any point, between the lines or multiple times in the middle of the single call. Nor of those should be considered as a delimiter between text blocks.

Also there is no such thing as "flush" syscall. There is only "write". Any flushing mechanics you may see in the field are heuristic guess at user space level how to optimize writing by buffering. As a result there is no such thing as a flush marker within the pipe communication. Keep in mind that ninja does not control such things, it acts as a transport system between the task process and client app.

> There are many programs which do care about the ordering

If you are talking about the order defined by the task processes then it can only be a result of defective design of such process. The only way order can be enforced on the reader side is by connecting both stdout and stderr on the writer side to the same pipe. Which does not make any sense and invalidates the whole idea of separate stderr usage. If process expects error output to be merged with the stdout then it should just write into stdout.

**Comment by jhasse** [comment]
> O_DIRECT is cannot be used, if you read the documentation about it you will see caveats about internal buffer size.

Do you mean "Writes of greater than PIPE_BUF bytes (see pipe(7)) will be split into multiple packets.  The constant PIPE_BUF is defined in <limits.h>."?

> Within those calls actual write could appear at any point, between the lines or multiple times in the middle of the single call. Nor of those should be considered as a delimiter between text blocks.

Most programs use line buffering by default so the write would happen after writing \n. Even if not, multiple writes will happen, I don't see the problem.

> As a result there is no such thing as a flush marker within the pipe communication.

The flush will result in a write, so there is a marker within the pipe communication. We just won't know if it was a flush a just the buffer being full, etc.

> If process expects error output to be merged with the stdout then it should just write into stdout.

I definitely agree. We don't life in this perfect world though ;)

Could ninja detect if its stderr is redirected to stdout?

**Comment by dendy** [comment]
> Do you mean "Writes of greater than PIPE_BUF bytes (see pipe(7)) will be split into multiple packets. The constant PIPE_BUF is defined in <limits.h>."?

PIPE_BUF is criminally low: 4096. But is does not matter for the ninja anyway. O_DIRECT could be used only when both reader and writer of the pipe follow predefined custom protocol. This is not the case for us, since writer is an arbitrary process.

> The flush will result in a write

That is not true, and even if it is it does it does not matter. write syscall does not mean bytes were actually written, if pipe buffer is full it writes arbitrary number of bytes and returns EAGAIN. Generally reader will receive number of bytes not correlated to the writes on the process side.

> Could ninja detect if its stderr is redirected to stdout?

Ninja itself does this redirection. Or I didn't get the question.

I honestly don't understand how the requirements you posted above were introduced:

1. > If this was to be fixed, it shouldn't be an option, but always handle stdout and stderr correctly.

Why this requirements exists? Normally new features first implemented in some experimental state, exposing the behavior change with argument switch, environment variable, configuration file, etc. Or even the build time option. Then it is developed, evaluated, concerns recorded and addressed, if any. Older behavior marked as deprecated and eventually removed. Or old and new remain coexist if both cases must be preserved.

If you have concerns that older behavior could be altered and that might be important, then requirement for mutually exclusive behavior is a road to nowhere.

2. > There are many programs which do care about the ordering

Where this statement comes from? I was trying to explain above that the nature of stdout and stderr design invalidates such cases, unless the task process if defected in first place. I don't believe such cases exists on practice and even if they do why ninja should care about them?

Consider next:

```
print('Processing foo...')
if error:
    print("Error!", file=sys.stderr)
else:
    print('Success')
```

The code above is brain damaged. I can't find where ninja is ever declared that the prints above will pop up in order. The fact that it might come in order of call sequence on the process side is pure coincidence.

On the other hand ninja does declare the behavior where stdout/stderr streams are not redirected to the stdout/stderr of the ninja itself. I assume such behavior of the 'make' is historical, where no system resources were available for buffering, nor most features were possible to implement without recursive Makefile, thus output of those sub Makefiles should be available immediately.

Ninja has no such issues and always does buffering, which is a good concept for the build system assembler. Separating stdout and stderr does not violate this principle. And this is the only thing ninja should consider to preserve from my point of view.

**Comment by jhasse** [comment]
> O_DIRECT could be used only when both reader and writer of the pipe follow predefined custom protocol.

Why is that?

> write syscall does not mean bytes were actually written

What happens with the bytes then if they aren't written?

> Ninja itself does this redirection. Or I didn't get the question.

I meant: Could ninja detect if it itself was called like this: `ninja 2>&1`?

> Then it is developed, evaluated, concerns recorded and addressed, if any.

And one concern is that with stderr not redirected to stdout lines that were previously in order aren't now.

> I can't find where ninja is ever declared that the prints above will pop up in order. The fact that it might come in order of call sequence on the process side is pure coincidence.

But they do come in order now that stderr is piped to stdout, do they not?

**Comment by dendy** [comment]
> Why is that?

Because that is the whole point of this flag, to receive complete packet in some format, which can be parsed as is, without seeking for begin/end and waiting for the remaining bytes in a stream. In our case there is no protocol and bytes will always come in arbitrary chunks.

> What happens with the bytes then if they aren't written?

Writer will know how many bytes were written and should try to send remaining again later. They might detect that with, for example, EPOLLOUT event of the epoll_ctl.

> I meant: Could ninja detect if it itself was called like this: ninja 2>&1?

I believe it is possible: https://man7.org/linux/man-pages/man2/kcmp.2.html

You mean ninja should try to detect that and switch to the old behavior? That would work, though I would prefer more explicit control over that by the users. If ninja command like is going to be modified anyway by the users, then it could be an argument to the CLI or environment variable.

> And one concern is that with stderr not redirected to stdout lines that were previously in order aren't now.

That is theoretical concern, it is unknown whether such issue exists in the field and how critical it is. I can't imagine this as a real issue.

Lets say there is a task app which does some job and reports normal logs and errors. There are two general cases:

1) This task app is general purpose one, it is also used outside of the ninja, for instance, in terminal or some background process. If so that task is already should follow common design principles of the stream separation. For example: C compiler.

2) This task is designed for and used just in ninja. In this case it does not need to write into stderr, because ninja redirects that into stdout anyway.

Both cases will not suffer from stdout and stderr separation. Existence of other cases are hypothetical.

> But they do come in order now that stderr is piped to stdout, do they not?

No, they are not in general. In fact even the below example would not produce sequential output in current ninja implementation:

```
# app does other prints into stdout and stderr, output pipe is almost full now

print("Status:")                          # writes "Sta", pipe is full now
if error:
    print("Error", file=sys.stderr)       # reader read some bytes, writes "Err" 
print("Done")
                                          # reader emptied the pipe
sys.stdout.flush()                        # writes "tus:Done"
sys.stderr.flush()                        # writes "or"
```

Ninja output:
```
StaErrtus:
Done
or
```

**Comment by jhasse** [comment]
Isn't that because of Python's buffering, not because of any pipe buffers?

**Comment by dendy** [comment]
Yes, that is because of the buffering in the user space. Any library does that, C++ stdlib, python, posix, Qt, etc. When app uses standard stream it assumes that byte queue of it is standalone. This is also true for pipes, sockets, files, and other types of streams.

**Comment by dendy** [comment]
@jhasse My proposal is to enable this feature in experimental mode. I am quite happy with the current patch (see link at the top).

With some caveats:
1. Default behavior should stay as in current implementation, to combine stdout and stderr.
2. Add environment variable to override the default behavior, e.g.: NINJA_USE_STDERR=1. This is needed in case it is convenient for the user to set behavior globally.
3. Add command line argument to also override the default behavior, which will have more priority than environment variable, e.g.: ninja --use-stderr=1. For instance, this is needed to use this feature in the IDE, without changing behavior globally.
4. (optionally) Add new binding, e.g. "use_stderr". This will allow to enable feature per individual command. Could be useful for generators.

Advertise this update in the new ninja version and collect feedback from the users about issues.

**Comment by dendy** [comment]
@jhasse Please let me know if the above is good for you, so I will proceed with the implementation.

**Comment by jhasse** [comment]
No, sorry.

**Comment by dendy** [comment]
Please elaborate. I explained everything in details in this thread. List your concerns so we could work on the plan.

**Comment by jhasse** [comment]
If I understood you correctly, stdout and stderr are just synchronized in the terminal by chance and the order may actually be totally random if you're unlucky. If this is true, people (like me) are using stderr wrong. Its usage should be discouraged as it's too easy to shoot yourself in the food. It isn't intuitive (i.e. the terminal mixes them by default). If someone wants "error" output separately he should explicitly write it to a file.

**Comment by sjg20** [comment]
> If I understood you correctly, stdout and stderr are just synchronized in the terminal by chance and the order may actually be totally random if you're unlucky. If this is true, people (like me) are using stderr wrong. Its usage should be discouraged as it's too easy to shoot yourself in the food. It isn't intuitive (i.e. the terminal mixes them by default). If someone wants "error" output separately he should explicitly write it to a file.

Excuse me but I think this is misconceived. The stdout output is not all that useful, to put it mildly. in fact I suppress it. It spits out a lot of stuff that indicates normal operation. For development and IDEs we want to see warnings and errors, nothing else. At present ninja provides no way to do this. The combining of stdout and stderr is strange, perhaps a Windows thing, but at the least ninja should provide a way to avoid this non-standard and (in my opinion) pointless output.

In general builds are mutil-threaded and the only thing matters is the filename, line number and message so I can work through those to fix the build. Even the compilation order is not that interesting.

So I propose changing the default to separate stdout and stderr. IMO that would be much more sensible, following make, gcc and the like.  If that is not acceptable, and we are happy to have this tool be different from everything else I have seen, then dendy@'s propose looks good to me :-)

In any case, please can we skip the circular discussion and implement *something* to fix this problem. We are currently using Zephyr with Cmake and engineers are asking about changing over to make it generate a Makefile instead...

**Comment by jhasse** [comment]
> At present ninja provides no way to do this.

Ninja might add `--quiet` in a future update so that you can hide status updates. Currently you can simulate that using `NINJA_STATUS` and the scripts above.

**Comment by dendy** [comment]
There are two types of terminals:
1. When terminal does not care about separating appearance of the stdout/stderr streams, so it combines them together by redirecting both streams into the same pipe. These are common console terminals. Same behavior when redirecting output with 2>&1. When write() syscall is used then output might appear in order. But any buffering on the writer side will cause output to be scrambled between stdout and stderr. You can never rely on what you see in the terminal as well as what you get in the 2>&1.
2. When terminal does care about stdout and stderr separation, for instance, to colorize the errors with red. Or hide/show one of the streams. In this case output will not be scrambled. All IDEs rely on this behavior to properly process command results.

> Its usage should be discouraged as it's too easy to shoot yourself in the foot

It is all fine with stderr stream as long as command which is writing into it is following the core logic, why those streams were invented years ago in first place: they are independent. There is no synchronization between them by design, nor any stream reader should rely on data order between streams.

Ninja cannot dictate policy to the executing commands how they should use stdout/stderr, nor to the caller how they should process output from the commands. Because nor of those is under control of the ninja, all those tools already exist.

> If someone wants "error" output separately he should explicitly write it to a file.

If someone wants stderr/stdout contents separately they should simply read appropriate stream. It is how all tools behave and there is nothing wrong with that.

Don't want to touch anyone's feelings, but no matter how I like the ideas implemented by ninja, for me personally ninja was totally useless until I implemented patch for stdout/stderr separation several years ago.

**Comment by sjg20** [comment]
+1 we have a hundred lines of Python code trying to clean up the ninja output and it is still not right. It messes up downstream things also - e.g. Zephyr is quite broken in terms of what is a warning/error and what is normal behaviour.

I wonder if the problem is that this tool came from Windows?

Do you have a link to the stdout/stderr patch? We could perhaps just use that.

**Comment by sjg20** [comment]
@dendy OK thank you. I filed https://issuetracker.google.com/187728036 so we'll see what we end up doing. Someone also suggesting switching Cmake to generate plain Makefiles.

**Comment by CarloWood** [comment]
This is a critical problem for me too - since @jhasse seems to refuse to add a simple option for it - where is the ninja fork that DOES implement this currently? Thanks.

**Comment by dendy** [comment]
@CarloWood https://github.com/dendy/ninja/tree/stderr

**Comment by orgads** [comment]
> @jhasse My proposal is to enable this feature in experimental mode. I am quite happy with the current patch (see link at the top).
> 
> With some caveats:
> 
> 1. Default behavior should stay as in current implementation, to combine stdout and stderr.
> 2. Add environment variable to override the default behavior, e.g.: NINJA_USE_STDERR=1. This is needed in case it is convenient for the user to set behavior globally.
> 3. Add command line argument to also override the default behavior, which will have more priority than environment variable, e.g.: ninja --use-stderr=1. For instance, this is needed to use this feature in the IDE, without changing behavior globally.
> 4. (optionally) Add new binding, e.g. "use_stderr". This will allow to enable feature per individual command. Could be useful for generators.
> 
> Advertise this update in the new ninja version and collect feedback from the users about issues.

@jhasse can you please explain why do you object to having this as opt-in, as suggested here? This is a real pain to many users.

**Comment by jhasse** [comment]
I'm no longer maintaining Ninja.

**Comment by jhasse** [comment]
misclicked, sorry

**Comment by orgads** [comment]
> I'm no longer maintaining Ninja.

Then who can approve this?

**Comment by kloczek** [comment]
Any update? 🤔

**Comment by dendy** [comment]
@jhasse Do you know who is the current maintainer of the ninja? We are still looking to accept this feature as optional, without changing the default behavior. Who could help to proceed with this?

**Comment by jhasse** [comment]
No one really.

> Ninja cannot dictate policy to the executing commands how they should use stdout/stderr, nor to the caller how they should process output from the commands.

Well, it can at least discourage the use of stderr by not providing an option to split stdout/stderr.

**Comment by sjg20** [comment]
+1 we need a proper stderr. This is such a pain. Why is this so controversial??

**Comment by dendy** [comment]
@jhasse Ninja is only executing other tools, and behavior of those tools how to produce data for stdout and stderr already well defined. There are billions of those tools, Ninja may not change they behavior.

In the case when parser of the Ninja output wants proper separation of stderr — there should be such an option.

My proposal above is to add this feature. And keep original behavior (combined output) as default. So everybody will win and be happy.

Have to rebase my patch onto recent master branch. Though first I will need your agreement in general to accept this feature.

**Comment by kloczek** [comment]
Gentle ping .. any update? 🤔

**Comment by tpimh** [comment]
@kloczek just use Daniel's fork, @jhasse is never going to merge this.

**Comment by kloczek** [comment]
> @kloczek just use Daniel's fork, @jhasse is never going to merge this.

OK so where is that fork and/or where I can find exact patch/commit? 😋

**Comment by kloczek** [comment]
I'm asking because I found only https://github.com/jhasse/ninja which seems is waaay older than what is here.

**Comment by kloczek** [comment]
https://github.com/jhasse/ninja seems is no longer maintained 🤔

**Comment by dendy** [comment]
@kloczek Look at the veeeery top message, you will find a link to my branch: https://github.com/dendy/ninja/tree/stderr
I will check if it needs to be updated by rebasing it onto the latest ninja mainline.

**Comment by kloczek** [comment]
> @kloczek Look at the veeeery top message, you will find a link to my branch: https://github.com/dendy/ninja/tree/stderr I will check if it needs to be updated by rebasing it onto the latest ninja mainline.

Can you resync your fork with ninja master? 🤔 
Currently that commit is not possible to apply on top of last version
```console
+ /usr/lib/rpm/rpmuncompress /home/tkloczko/rpmbuild/SOURCES/ninja-build-Add-option-to-split-stdout-and-stderr.patch
6 out of 8 hunks FAILED -- saving rejects to file src/build.cc.rej
2 out of 3 hunks FAILED -- saving rejects to file src/build.h.rej
1 out of 11 hunks FAILED -- saving rejects to file src/subprocess-posix.cc.rej
```

**Comment by Ext3h** [comment]
Not really happy  with the approach of passing through `stdout` / `stderr` either. It's an option - sure - but it's not even close to actually fixing the issue at hand, only introducing a new bug. It's doing the same stuff wrong as `make`s `--line-buffer` does.

The problem we have, is that called applications end up assuming that full buffering was permissible, even when it clearly wasn't.

> FWIW ninja is corrupting the output from Doxygen (via CMake's doxygen_add_docs()). The issue doesn't happen when using GNU Make (or Doxygen directly).

E.g. in this case, the patch with split `stderr` only looks like it does something sensible at first glance, because when forwarding `stderr` the terminal has enforced line buffering for `ninja`s output buffers, but the output from the original application still had a fully buffered `stdout` due to being attached to a pipe rather than a pty.

That's not working properly at all with applications that split time correlated output onto different output streams, such as Doxygen which is splitting progress and warning messages. You still get warning messages completely out-of-sync with the progress messages they should have been embedded in between.

Doxygen actually has a `-b` command line argument to ***force*** un-buffered output in order to work correctly when embedded into other applications.

> Actually I am back to using ninja. By prefixing my commands with `stdbuf -oL -eL` (from GNU coreutils) everything works as it should.

That did do something sensible. Partially. The hack of preloading `libstdbuf` unfortunately isn't exactly portable.

In case of Doxygen, the actual sane solution is to specifiy `WARN_LOGFILE = stdout` or alike - telling Doxygen that it should stop splitting messages between stdout and stderr. (See https://gitlab.kitware.com/cmake/cmake/-/issues/21692 - it's not working with CMake either, unfortunately.)

**Comment by CarloWood** [comment]
I just wrote this at the time and never looked back:
```
#include <signal.h>
#include <stdlib.h>
#include <sysexits.h>
#include <unistd.h>
#include <stdio.h>
#include <pty.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>

pid_t child = 0;

void sighandler(int signum)
{
  if (child > 0)
  {
    killpg(child, signum);
    exit(signum);
  }
}

int last_is_esc_K = 0;

char const* find_eom(char const* start, char const* end)
{
  for (char const* ptr = start; ptr < end; ++ptr)
  {
    if (*ptr == '\033')
    {
      if (ptr < end - 2 && ptr[1] == '[' && ptr[2] == 'K')
      {
        last_is_esc_K = 1;
        return ptr + 3;
      }
    }
    if (*ptr == '\n')
    {
      last_is_esc_K = 0;
      return ptr + 1;
    }
  }
  return NULL;
}

int errors = 0;

void handle_msg(char const* start, size_t len)
{
  char const* p = start;
  char const* e = start + len;
  int print_stdout = 0;
  if (len >= 6)
  {
    if (strncmp(start, "ninja:", 6) == 0)
      print_stdout = 1;
    else if (len >= 17 && strncmp(start, "\e[31mFAILED: \e[0m", 17) == 0)
      ++errors;
    else
    {
      // \r[12/42]
      //  ^  ^
      //  |  |
      //  p  +3
      if (p[0] == '\r' && p[1] == '[' && isdigit(p[2]))
      {
        p += 3;
        while (p < e && isdigit(*p))
          ++p;
        // \r[12/42]
        //      ^ ^
        //      | |
        //      p +2
        if (p + 2 < e)
        {
          if (p[0] == '/' && isdigit(p[1]))
          {
            p += 2;
            while (p < e && isdigit(*p))
              ++p;
            // \r[12/42]
            //         ^
            //         |
            //         p
            if (p < e && *p == ']')
              print_stdout = 1;
          }
        }
      }
    }
  }
  else
  {
    while (!print_stdout && p < e && (*p == '\r' || *p == '\n'))
    {
      if (*p == '\n')
        print_stdout = 1;
      ++p;
    }
  }
  if (print_stdout)
  {
    fflush(stderr);
    if (write(STDOUT_FILENO, start, len) != len)
    {
      perror("failed to write to stdout");
      exit(EX_OSERR);
    }
  }
  else if (errors <= 1)
  {
    fflush(stdout);
    if (write(STDERR_FILENO, start, len) != len)
    {
      perror("failed to write to stderr");
      exit(EX_OSERR);
    }
  }
#if 0
  else
  {
    fprintf(stderr, "Suppressed:\n");
    fflush(stderr);
    if (write(STDERR_FILENO, start, len) != len)
    {
      perror("failed to write to stderr");
      exit(EX_OSERR);
    }
  }
#endif
}

// Run a command in a pty.
// Usage: /path/to/this/binary command to run.
int main(int argc, char *argv[])
{
  if (argc < 2)
    return EX_USAGE;

  int master;
  child = forkpty(&master, NULL, NULL, NULL);

  if (child == -1)
  {
    perror("failed to fork pty");
    return EX_OSERR;
  }

  if (child == 0)
  {
    // We're in the child process, so replace it with the command.
    execvp(argv[1], argv + 1);
    perror("failed to execute command");
    return EX_OSERR;
  }

  // Trap kill signals and forward them to child process.
  signal(SIGHUP, sighandler);
  signal(SIGINT, sighandler);
  signal(SIGTERM, sighandler);

  int const buf_size = 1024;
  char buf[buf_size + 1];
  buf[buf_size] = 0;
  char record[buf_size];
  fd_set fds;
  char* pptr = buf;                             // Begin writing at the start of the buffer.
  char const* const epptr = buf + buf_size;     // One past the end of the buffer.
  char const* egptr = pptr;                     // The buffer is empty.
  char const* gptr = egptr;                     // There is nothing to get.

  // Forward the output continuously.
  while (1)
  {
    FD_ZERO(&fds);
    FD_SET(master, &fds);

    if (select(master + 1, &fds, NULL, NULL, NULL) > 0 && FD_ISSET(master, &fds))
    {
      ssize_t bytes_read = read(master, pptr, epptr - pptr);
      if (bytes_read == 0 || (bytes_read == -1 && errno == EIO))
      {
        if (last_is_esc_K)
          printf("\r\n");
        return EXIT_SUCCESS;
      }
      if (bytes_read < 0)
      {
        perror("error reading input");
        return EX_OSERR;
      }

      pptr += bytes_read;
      egptr += bytes_read;

      // Parse the input.
      //                                                           pptr  epptr
      //                                                            |      |
      //                                                            v      v
      // [<old message><new message1><new message2><partial message><junk>]
      // ^             ^                                            ^
      // |             |                                            |
      //buf           gptr                                        egptr
      for (;;)
      {
        char const* eom = find_eom(gptr, egptr);
        if (eom)
        {
          handle_msg(gptr, eom - gptr);
          gptr = eom;
          continue;
        }
        else if (pptr == epptr)
        {
          eom = egptr;
          handle_msg(gptr, egptr - gptr);
          gptr = egptr;
        }
        break;
      }
      if (gptr != buf)
      {
        size_t size = egptr - gptr;
        size_t shift = gptr - buf;
        if (shift > 0 && size > 0)
          memmove(buf, gptr, size);
        gptr = buf;
        egptr = gptr + size;
        pptr -= shift;
      }
    }
  }
}
```

This reads from stdin, and fakes being a tty - so that the program that is passed doesn't buffer.
It then splits the input, on a line by line basis, to stdout or stderr, preserving the order by flushing whenever it switches
between the two. The decision on whether to write stdout or stderr is made based on the content
of the line.

I'm using it like this:
```
                    if [[ $CMAKE_CONFIGURE_OPTIONS == *"-GNinja"* ]]; then
                        /usr/local/bin/ninja-filter cmake --build "$BUILDDIR" --config "$CMAKE_CONFIG" --parallel $CPUS -- $@ 2> >(awk 'BEGIN { have_match = 0 } /operator<</ { if (have_match == 2) { have_match = 1; next }} /^\/usr\/.*(no known conversion from|candidate template ignored: |template argument deduction\/substitution failed:)/ { have_match = 2; keep = $0; next } { if (have_match == 2) { print keep; have_match = 0 } if (have_match == 0) print; else { have_match = 0 }}' | sed -e 's%'"$REPOBASE"'/%%;s/symbolic:://g' 1>&2);
                    else
                        cmake --build "$BUILDDIR" --config "$CMAKE_CONFIG" --parallel $CPUS -- $@;
                    fi;
```
from a bash script.

**Comment by dendy** [comment]
@kloczek Apologies for the late response, just rebased stderr branch onto recent master. Please check it out.

**Comment by macdew** [comment]
I'd like to add a "me too" here as well, the merging of the output streams completely destroys our handling of the output of the applications that ninja is calling.  We've now switched back to Make due to this.  I'm very sad to read as well that Ninja is no longer being maintained, it was a good build engine for a while.
